app/src/main/java/com/simplecity/amp_library/playback/MusicService.java

```diff
@@ -1696,57 +1613,48 @@ public void makeTrackShuffleList() {
         }
     }
 
-    public void makeAlbumShuffleList() {
+    public void makeAlbumShuffleList(boolean startAtFront) {
         synchronized (this) {
-            if (albumShuffleList.size() > 0) {
-                albumShuffleList = new ArrayList<>();
-            }
-
             if (playlist == null || playlist.isEmpty()) {
                 return;
             }
 
-            Map<Album, List<Song>> trackShuffleListAlbumMap = new HashMap<>();
-
-            for (Song song : trackShuffleList) {
-                Album containingAlbum = song.getAlbum();
-                if (!trackShuffleListAlbumMap.containsKey(containingAlbum)) {
-                    trackShuffleListAlbumMap.put(containingAlbum, new ArrayList<>());
-                }
-                trackShuffleListAlbumMap.get(containingAlbum).add(song);
+            if (albumShuffleList.size() > 0) {
+                albumShuffleList = new ArrayList<>();
             }
 
-            Song currentSong = null;
-            if (playPos >= 0 && playPos < playlist.size()) {
-                currentSong = trackShuffleList.get(playPos);
-            }
+            // We are shuffling the current queue, so current song is within the list.
+            // trackShuffleList is already shuffled, so we don't need to preshuffle.
+            albumShuffleList.addAll(albumShuffleSongs(
+                    trackShuffleList.isEmpty() ? playlist : trackShuffleList,
+                    false, false));
 
-            if (currentSong != null) {
-                trackShuffleListAlbumMap.get(currentSong.getAlbum()).remove(currentSong);
-                albumShuffleList.add(currentSong);
-                albumShuffleList.addAll(trackShuffleListAlbumMap.get(currentSong.getAlbum()));
-                trackShuffleListAlbumMap.remove(currentSong.getAlbum());
+            if (startAtFront) {
+                playPos = 0;
+            } else  {
+                playPos = currentSong.getAlbum().getSongsSingle().blockingGet().indexOf(currentSong);
             }
+        }
+    }
 
-            List<Album> trackShuffleListAlbums = new ArrayList<>(trackShuffleListAlbumMap.keySet());
-            Collections.shuffle(trackShuffleListAlbums);
-
-            for (Album album : trackShuffleListAlbums) {
-                albumShuffleList.addAll(trackShuffleListAlbumMap.get(album));
-            }
+    private List<Song> albumQueueIntersect(Album album, List<Song> albumSongsInQueue) {
+        List<Song> albumSongsInOrder = new ArrayList<>(album.getSongsSingle().blockingGet());
```

Comment:
Another `blockingGet()` call which should be refactored, or we'll block the entire thread. I would make `albumQueueIntersect` return a `Single<List<Song>>` instead of `List<Song>`.